server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java

```diff
@@ -168,8 +186,64 @@ private void doRedirectFailures() {
         )) {
             return;
         }
-        Map<ShardId, List<BulkItemRequest>> requestsByShard = drainAndGroupRedirectsByShards(clusterState);
-        executeBulkRequestsByShard(requestsByShard, clusterState, this::completeBulkOperation);
+        Runnable executeRedirectRequests = () -> {
+            // Get new cluster state that includes any potential failure store rollovers.
+            var rolledOverState = observer.setAndGetObservedState();
```

Comment:
Is there a chance we could end up with a new cluster block here? I don't know if we should do another round of ensuring blocks are handled or not.

Reply:
I'm not too familiar with cluster blocks, but wouldn't it always be possible to get a cluster block right after obtaining a state that doesn't have one? Our `ClusterStateObserver` waiting for and obtaining a cluster state that doesn't have a block, doesn't mean that the cluster can't get blocked right after (when we start grouping the request by shard or start sending rollover requests), right?

Reply:
I think this method just returns the last observed value right? It seems to me that it just retrieves the latest one.

https://github.com/elastic/elasticsearch/blob/08c7b655a2dff7f0c66db5247a7f7fd0dd66fa88/server/src/main/java/org/elasticsearch/cluster/ClusterStateObserver.java#L92-L99

Reply:
Yeah I agree with what @jbaiera was saying. My point was that I don't really know what the value would be of doing another check, as even if we do a check, the cluster could become blocked 1ms after we do the check, and we'd proceed rolling over the lazy rollover, right?

Reply:
> the cluster could become blocked 1ms after we do the check, and we'd proceed rolling over the lazy rollover

This is the reason I'm kind of on the fence about doing anything about it. I wanted to bring it up just to make sure we were aware about it, but after taking some time to dig through the cluster blocks some more I don't think I have a preference for making a change here at the moment.

Reply:
Thanks for bringing it up and doing some digging, @jbaiera!